src/consensus/versionbits.h

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) 2016 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_CONSENSUS_VERSIONBITS
+#define BITCOIN_CONSENSUS_VERSIONBITS
+
+#include "chain.h"
+#include <map>
+
+namespace Consensus
+{
+
+/** What block version to use for new blocks (pre versionbits) */
+static const int32_t VERSIONBITS_LAST_OLD_BLOCK_VERSION = 4;
+/** What bits to set in version for versionbits blocks */
+static const int32_t VERSIONBITS_TOP_BITS = 0x20000000UL;
+/** What bitmask determines whether versionbits is in use */
+static const int32_t VERSIONBITS_TOP_MASK = 0x60000000UL;
+
+/**
+ * Abstract class that implements BIP9-style threshold logic, and caches results.
+ */
+class AbstractThresholdConditionChecker {
+protected:
+    virtual bool Condition(const CBlockIndex* pindex, const Params& params) =0;
+    virtual int64_t BeginTime(const Params& params) =0;
+    virtual int64_t EndTime(const Params& params) =0;
+    virtual int Period(const Params& params) =0;
+    virtual int Threshold(const Params& params) =0;
+
+    enum State {
+        UNDEFINED,
+        DEFINED,
```

Comment:
Bike-shedding: I'm doing s/UNDEFINED/DEFINED s/DEFINED/STARTED isntead, which I think it's more clear:
If it's in the chainparams is defined, it just hasn't started yet. Undefined/unkown deployments are those which you don't have in your chainparams.